New posted card view enhancements

hunbernd opened the pull request.

defnax committed:
Added new Card View for Posted

defnax committed:
fixed loading issue for card view

hunbernd committed:
Really fixed loading issue for card view

hunbernd committed:
Image optimizer can output byte array

hunbernd committed:
Merge branch 'upstream' into foreign/New-Posted-Card-View

hunbernd committed:
Removed redundant title inputbox

hunbernd committed:
Auto resize image to fit into 34kB

hunbernd committed:
Fixed bug preventing hashing files from file paths with special characters

hunbernd committed:
Link the original image

defnax committed:
Fixed Card View display by anmo

hunbernd committed:
Do not allow posting too large messages

hunbernd committed:
Merge defnax's branch

defnax commented:
@csoler you can check it soon?

csoler commented on retroshare-gui/src/gui/Posted/PostedCreatePostDialog.cpp:
```diff
@@ -85,37 +101,27 @@ void PostedCreatePostDialog::createPost()
-		buffer.open(QIODevice::WriteOnly);
-		picture.save(&buffer, "PNG"); // writes image into ba in PNG format
-		post.mImage.copy((uint8_t *) ba.data(), ba.size());
+	int msgsize = post.mLink.length() + post.mMeta.mMsgName.length() + post.mNotes.length() + imagebytes.size();
```
When the msg is too large, it would be useful to allow the user to temporarily share the image and replace it with a retroshare link.

csoler replied:
(this is suggestion for later improvement. Not needed for this particular  PR).

csoler commented on retroshare-gui/src/gui/Posted/PostedCreatePostDialog.cpp:
```diff
@@ -124,17 +130,62 @@ void PostedCreatePostDialog::createPost()
-	picture = img;
+	//Do we need to hash the image?
+	QMessageBox::StandardButton answer;
+	answer = QMessageBox::question(this, tr("Post image"), tr("Do you want to share and link the original image?"), QMessageBox::Yes|QMessageBox::No);
```
Great! I didn't see it before. So my previous comment is irrelevant.

csoler commented on retroshare-gui/src/gui/Posted/PostedListWidget.cpp:
```diff
@@ -534,11 +630,53 @@ void PostedListWidget::applyRanking()
 		++counter;
 	}
 
+	// Card View
```
In cardview mode, there's a variable gap between items in the list. Is it possible to remove this gap, so that all items whatever their size, get properly adjusted?

csoler commented on retroshare-gui/src/util/imageutil.cpp:
```diff
@@ -137,34 +137,40 @@ bool ImageUtil::optimizeSize(QString &html, const QImage& original, QImage &opti
+{
+	QByteArray bytearray;
+	if(maxBytes > 0){
+		maxBytes = maxBytes * 3/4 - 50; //base64 and html stuff
```
What happens if someone calls this with maxBytes <= 64 ? I think you should add a test here to make sure this doesn't cause some unwanted behavior.

defnax committed:
Added anmos changes & own improvements

Added the changes from anmo
Added to store/load last used identity for posted composer
Added remove button for remove the attached image
Fixed to hide the Notes button when there is not text content (PosteItem)
Fixed the ui margins better look

felisucoibi commented:
+1

defnax committed:
forget to commit this

hunbernd committed:
Merge remote-tracking branch 'remotes/defnax/New-Posted-Card-View'

hunbernd committed:
Handle an edge case in imageutil

hunbernd committed:
Rise post size

defnax committed:
Merge pull request #43 from hunbernd/New-Posted-Card-View

New posted card view

defnax committed:
Added gif & webp for RichtextEditor

* Added gif & webp for RichtextEditor picture attachments
* Attempt to fix attached picture layout issue on Card View

defnax committed:
fixing margins & hide notes field when its empty

defnax committed:
set default picture dimensions to 640*480 like same as embeded html image

* set default picture dimensions to 640*480 like same as embeded html image

defnax committed:
Improved the Image attach Page

defnax commented:
@csoler its done now, please check

hunbernd committed:
Merge upstream

hunbernd committed:
Fixes the build problem with the new notification system.
PostedCardView uses the same identifier as PostedItem.

hunbernd commented:
I am not sure if it is OK to use the same identifier in both places, but I cannot test it because I don't get notifications from posted due to an other bug not from this PR.

defnax committed:
Merge pull request #45 from hunbernd/New-Posted-Card-View

New posted card view

defnax committed:
Added Picture View Dialog for open the picture in a external View

* Added Picture View Dialog for open the picture in a external View
* Added a clickable Label for Classic Views Thumbnail label to open Picture via one click
* Imroved the Classic & Card views dark stylesheets for better look and feel

hunbernd committed:
Merge upstream

defnax committed:
added delete on close

defnax committed:
added temporary solution to not stretch the pixmap out of the window

defnax committed:
attempt to fix issues with pictures which width are bigger to fit on Label

defnax committed:
corrected to use the right value

defnax committed:
get the share button usable for copy the rs link

defnax committed:
fixing stylesheets for clickable label

defnax committed:
set stylesheet for the new label & removed old stylesheets from the channel item

defnax committed:
added to check pixmap width & height before scale, only scale when needed

csoler commented:
Hi, I fixed the other bug, so you should be able to test now. Also this PR has some conflicts now.

hunbernd committed:
Merge remote-tracking branch 'remotes/origin/pr/2' into feature/New-Posted-Card-View

defnax committed:
attempt to improve

defnax committed:
Added the new class for resize the image if the window is resized

felisucoibi commented:
+1

hunbernd committed:
Merge remote-tracking branch 'remotes/origin/pr/2' into feature/New-Posted-Card-View

hunbernd committed:
Make pixmap resizing label work

hunbernd committed:
Resize photoview to the size of embedded image

hunbernd committed:
Removed size limit inside image viewer, because QT already limits the window size to the 2/3 of screen size

hunbernd committed:
Merge upstream

defnax committed:
Merge pull request #48 from hunbernd/New-Posted-Card-View

New posted card view

defnax commented:
> Hi, I fixed the other bug, so you should be able to test now. Also this PR has some conflicts now.

but the postedlist still doesnt yet update when i post a new post, i need to select other topic to get updated. or this issue is only on win

defnax committed:
adding back spacer for PhotoView & fixed issue with centering

defnax commented:
i have fixed the issue with centering the picture @hunbernd 
https://github.com/defnax/RetroShare/raw/bd18f572c9b0683a7d088336a10f12c5b461a897/retroshare-gui/src/gui/Posted/PhotoView.ui

or i make new pr after this is merged for fix the Photo View centering

defnax committed:
Enabled Min & Max feature for Photo View Dialog

csoler commented:
you're either missing a file, or have a case problem in this line:
 #include "util/aspectratiopixmaplabel.h"
that's why the Travis checks fail.

defnax commented:
> you're either missing a file, or have a case problem in this line:
> #include "util/aspectratiopixmaplabel.h"
> that's why the Travis checks fail.

i think its used wrong naming on PhotoView the class, normal its this right "util/AspectRatioPixmapLabel.h"

@hunbernd i have created new pr for the fixes
https://github.com/hunbernd/RetroShare/pull/3

defnax committed:
fixed include

csoler commented:
the problem is in PhotoView.ui, when you use the AspectRatioPixmapLabel class, the .h does not have the right case. You need to either wix it in the .ui or use qtcreator and fix the location of the file.

csoler commented:
Nice job!
1 - it seems that you removed the expand/collapse button. Is it on purpose?
2 - loading all widgets at once proves to be a bit slow, probably due to many calls to GXS internals. I suggest that you only load them when they are displayed. That's what we do already in channels. You may take inspiration from the code there.

defnax commented:
> Nice job!
> 1 - it seems that you removed the expand/collapse button. Is it on purpose?

No its not removed & i dont want it to remove.
PhotoView is a Optional View to view the Photo in the right dimensions like how on reddit
as external viewer.
and the expand button will stay to view on item.

In future i thinked we need for Photoview a feature to go back < & > Forward to view all posted pics from a Topic & maybe a preview some last pictures list...

![image](https://user-images.githubusercontent.com/9952056/73123633-38445980-3f92-11ea-8b22-e8ca3e8dda4d.png)


> 2 - loading all widgets at once proves to be a bit slow, probably due to many calls to GXS internals. I suggest that you only load them when they are displayed. That's what we do already in channels. You may take inspiration from the code there.

yes i know this, i have no time to improve that loading thing, more posts it takes more time to load..
last year i has experimented to cloned the channel method on posted, it was instabile then i deleted my code.
